### PR TITLE
style: Some animation cleanups and fixes.

### LINKED ISSUE #20116: Servo's transitions code stopped making sense after #15900.
In particular, what used to update the _new_ style (as expected, since that's what you're animating), started updating the old style:

```rust
       if let Some(ref mut old) = *old_values {
           self.update_animations_for_cascade(shared_context, old,
                                              possibly_expired_animations);
       }
```

Which doesn't really make sense.

### LINKED ISSUE #20731: Keyframe animations are not stopped/removed properly
Consider the following test document:
```html
<!DOCTYPE html>
<html>
<style>
    @keyframes spin {
        0% {
            transform: rotate(0deg);
        }
        100% {
            transform: rotate(360deg);
        }
    }

    .spin {
        animation: spin 2s infinite linear;
    }

    #box {
        border: 3px solid black;
        padding: 1em;
    }
</style>
<script>
    // Stop the animation after 2 seconds.
    window.setTimeout(() => {
        document.getElementById("box").classList.remove("spin");
    }, 2000);
</script>

<body>
    <h1 id="box" class="spin">Hello</h1>
</body>

</html>
```

This spins the `h1` for 2 seconds then stops animating. Visually, everything looks ok (though there is some jank in the animation), but if your run with `RUST_LOG=debug` you will notice that there is a constant flux of ` WARN 2018-05-01T18:39:03Z: style::animation: update_style_for_animation: Animation Atom('spin' type=inline) not found in style` logs. It seems to be that the animation is still in a `Running` state, even though no element is using it anymore. This doesn't happen at all if we don't start the animation by not setting the `spin` class in the first place.

On desktop the performance may not seem to drop much, but it's really bad on a mid-range mobile (32bits quad core arm at 1GHz).

### PR DESCRIPTION
**Pull request by emilio**
The transitions code is still terribly broken, but I ran out of time fixing it. We have nothing that stops transitions, which is just plain wrong. Most of this code should probably be rewritten, since with the current setup is pretty hard to get it right. Anyway...

Fixes #20731.
Fixes #20116.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20757)

### CODE REVIEW

**Comment by CYBAI** [comment]
In `linux-rel-css`:
```
{"status": "FAIL", "group": "default", "message": "assert_in_array: color is green after transition runs value \"rgb(0, 0, 255)\" not in array [\"rgb(0, 128, 0)\", \"rgba(0, 128, 0, 1)\"]", "stack": "@http://web-platform.test:8000/css/css-variables/variable-transitions-transition-property-all-before-value.html:59:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1490:20\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1530:17\n", "subtest": "Verify substituted color value after transition", "test": "/css/css-variables/variable-transitions-transition-property-all-before-value.html", "line": 40081, "action": "test_result", "expected": "PASS"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "font-style animation", "test": "/css/css-fonts/variations/font-style-interpolation.html", "line": 60956, "action": "test_result", "expected": "TIMEOUT"}
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "font-style transition", "test": "/css/css-fonts/variations/font-style-interpolation.html", "line": 60957, "action": "test_result", "expected": "PASS"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "font-weight animation", "test": "/css/css-fonts/variations/font-weight-interpolation.html", "line": 104452, "action": "test_result", "expected": "TIMEOUT"}
{"status": "FAIL", "group": "default", "message": "assert_in_array: color is green after transition runs value \"rgb(0, 0, 255)\" not in array [\"rgb(0, 128, 0)\", \"rgba(0, 128, 0, 1)\"]", "stack": "@http://web-platform.test:8000/css/css-variables/variable-transitions-value-before-transition-property-all.html:58:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1490:20\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1530:17\n", "subtest": "Verify substituted color value after transition", "test": "/css/css-variables/variable-transitions-value-before-transition-property-all.html", "line": 115469, "action": "test_result", "expected": "PASS"}
```

In `mac-rel-wpt4`:
```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "font-style animation", "test": "/css/css-fonts/variations/font-style-interpolation.html", "line": 22039, "action": "test_result", "expected": "TIMEOUT"}
{"status": "TIMEOUT", "group": "default", "message": "Test timed out", "stack": null, "subtest": "font-style transition", "test": "/css/css-fonts/variations/font-style-interpolation.html", "line": 22040, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "Overall test", "test": "/_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/swizzlers/swizzlers_017_to_024.html", "line": 90995, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/swizzlers/swizzlers_017_to_024.html", "line": 91012, "action": "test_result", "expected": "OK"}
```

File: components/style/properties/helpers/animated_properties.mako.rs [hunk 1]
```diff
@@ -213,17 +213,22 @@ pub enum AnimatedProperty {
...
+    /// Get the id of the property we're animating.
+    pub fn id(&self) -> LonghandId {
         match *self {
             % for prop in data.longhands:
-                % if prop.animatable:
-                    AnimatedProperty::${prop.camel_case}(..) => "${prop.name}",
-                % endif
+            % if prop.animatable:
+            AnimatedProperty::${prop.camel_case}(..) => LonghandId::${prop.camel_case},
+            % endif
```
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: I wonder if `AnimatedProperty` should use the same trick as `AnimationValue` to just read the discriminant value as the `LonghandId` result.

**Comment by fabricedesre** [comment]
@emilio do you need help pushing that to the finish line?

**Comment by emilio** [comment]
I just rebased it, I don't think the failures will go away, though I think we should land this regardless because our transitions code makes no sense and these tests are just passing by chance.

**Comment by emilio** [comment]
@bors-servo try

 * Now building.

**Comment by jdm** [comment]
I would accept that argument.

**Comment by emilio** [comment]
@bors-servo try

 * I think this should be ready for review.

**Comment by jdm** [comment]
The results for these tests appear stable on linux and mac:
* /css/css-transitions/properties-value-003.html
* /css/css-transitions/properties-value-001.html
* /css/css-transitions/properties-value-inherit-002.html
* /css/css-transitions/properties-value-implicit-001.html

**Comment by emilio** [comment]
Hmm, that's somewhat surprising, because the only difference with the previous try run (which didn't have that test-case) is the test expectations...

**Pull request merged**